bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/discovery/RokuDiscoveryService.java

```diff
@@ -0,0 +1,267 @@
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link RokuDiscoveryService} is responsible for discovery of Roku devices on the local network
+ *
+ * @author William Welliver - Initial contribution
+ * @author Dan Cunningham - Refactoring and Improvements
+ * @author Michael Lobstein - Modified for Roku binding
+ */
+
+@NonNullByDefault
+@Component(service = DiscoveryService.class, configurationPid = "discovery.roku")
+public class RokuDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(RokuDiscoveryService.class);
+    private static final String ROKU_DISCOVERY_MESSAGE = "M-SEARCH * HTTP/1.1\r\n" + "Host: 239.255.255.250:1900\r\n"
+            + "Man: \"ssdp:discover\"\r\n" + "ST: roku:ecp\r\n" + "\r\n";
+
+    private static final Pattern USN_PATTERN = Pattern.compile("^(uuid:roku:)?ecp:([0-9a-zA-Z]{1,16})");
+
+    private static final Pattern IP_HOST_PATTERN = Pattern
+            .compile("([0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}):([0-9]{1,5})");
+
+    private static final String ROKU_SSDP_MATCH = "uuid:roku:ecp";
+    private static final int BACKGROUND_SCAN_INTERVAL_SECONDS = 300;
+
+    private final HttpClient httpClient;
+
+    private @Nullable ScheduledFuture<?> scheduledFuture;
+
+    @Activate
+    public RokuDiscoveryService(final @Reference HttpClientFactory httpClientFactory) {
+        super(SUPPORTED_THING_TYPES_UIDS, 30, true);
+        this.httpClient = httpClientFactory.getCommonHttpClient();
+    }
+
+    @Override
+    public void startBackgroundDiscovery() {
+        stopBackgroundDiscovery();
+        scheduledFuture = scheduler.scheduleWithFixedDelay(this::doNetworkScan, 0, BACKGROUND_SCAN_INTERVAL_SECONDS,
+                TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void stopBackgroundDiscovery() {
+        ScheduledFuture<?> scheduledFuture = this.scheduledFuture;
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+        }
+        this.scheduledFuture = null;
+    }
+
+    @Override
+    public void startScan() {
+        doNetworkScan();
+    }
+
+    /**
+     * Enumerate all network interfaces, send the discovery broadcast and process responses.
+     *
+     */
+    private synchronized void doNetworkScan() {
+        try {
+            Enumeration<NetworkInterface> nets = NetworkInterface.getNetworkInterfaces();
+            while (nets.hasMoreElements()) {
+                NetworkInterface ni = nets.nextElement();
+                DatagramSocket socket = sendDiscoveryBroacast(ni);
+                if (socket != null) {
+                    scanResposesForKeywords(socket);
+                }
```

Comment:
If you put the `DatagramSocket` into a `try-with-resources` you can omit all the calls to `close`.
```suggestion
                try (DatagramSocket socket = sendDiscoveryBroacast(ni)) {
                    if (socket != null) {
                        scanResposesForKeywords(socket);
                    }
                }
```